### PR TITLE
[New protocol] Add WMI support

### PR DESCRIPTION
**Pull request by XiaoliChan**
WMI brute force (or maybe call it RPC brute force?), for more details: [#657](https://github.com/Porchetta-Industries/CrackMapExec/pull/657)

Everythings work now :)

SMB enumeration is gone, I pick the functions from `DumpNTLMInfo.py` to implement host info enumeration.

- Unprivileged users
image

- Admin privilege users
image

- Kerberos login:
image 

- Also, I try this in a real-world attack (445 blocked, only accept 135 traffic)
image

### CODE REVIEW

**Comment by XiaoliChan** [comment]
@mpgn I can confirm everything works now, please give a review :)

**Comment by mpgn** [comment]
okay ^^

File: cme/protocols/wmi/wmiexec_regout.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        return data
+
+    def execute_remote(self, data):
+        # Save result as txt file
+        resultTXT = "C:\\windows\\temp\\" + str(uuid.uuid4()) + ".txt"
+        self.logger.success("Executing command: \" %s \""%data)
+        data = data + " > " + resultTXT
+        command = self.__shell + self.encodeCommand(data)
+        self.__win32Process.Create(command, self.__pwd, None)
+        self.logger.highlight("[+] Waiting {}s for command completely executed.".format(self.interval_time))
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:46]: self.logger.info

File: cme/protocols/wmi/wmiexec_regout.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        time.sleep(self.interval_time)
+        
+        # Convert result to base64 strings
+        self.logger.highlight("[+] Save file to: " + resultTXT)
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:50]: self.logger.info

File: cme/protocols/wmi/wmiexec_regout.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        keyName = str(uuid.uuid4())
+        data = """[convert]::ToBase64String((Get-Content -path %s -Encoding byte)) | set-content -path C:\\windows\\temp\\%s.txt -force | Out-Null"""%(resultTXT,keyName)
+        command = self.__shell + self.encodeCommand(data)
+        self.__win32Process.Create(command, self.__pwd, None)
+        self.logger.highlight("[+] Waiting {}s for command completely executed.".format(self.interval_time))
+        time.sleep(self.interval_time)
+        
+        # Add base64 strings to registry
+        registry_Path = "HKLM:\\Software\\Classes\\hello\\"
+        self.logger.highlight("[+] Adding base64 strings to registry, path: %s, keyname: %s"%(registry_Path,keyName))
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:60]: self.logger.info

File: cme/protocols/wmi/wmiexec_regout.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        data = """New-Item %s -Force; New-ItemProperty -Path %s -Name %s -Value (get-content -path C:\\windows\\temp\\%s.txt) -PropertyType string -Force | Out-Null"""%(registry_Path,registry_Path,keyName,keyName)
+        command = self.__shell + self.encodeCommand(data)
+        self.__win32Process.Create(command, self.__pwd, None)
+        self.logger.highlight("[+] Waiting {}s for command completely executed.".format(self.interval_time))
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:64]: self.logger.info

File: cme/protocols/wmi/wmiexec_regout.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        time.sleep(self.interval_time)
+        
+        # Remove temp file
+        self.logger.highlight("[+] Remove temporary files")
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:68]: self.logger.info

File: cme/protocols/wmi/wmiexec_regout.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        
+        # Query result through WQL syntax
+        self.queryWQL(keyName)
+
+    def queryWQL(self, keyName):
+        namespace = '//%s/root/default' % self.address
+        descriptor, _ = self.iWbemServices.GetObject('StdRegProv')
+        descriptor = descriptor.SpawnInstance()
+        retVal = descriptor.GetStringValue(2147483650,'SOFTWARE\\classes\\hello', keyName)
+        self.logger.highlight("[+] Get result:")
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:81]: self.logger.info

File: cme/protocols/wmi/wmiexec_regout.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        result = retVal.sValue
+        self.logger.highlight(base64.b64decode(result).decode('utf-16le'))
+        self.logger.highlight("[+] Remove temporary registry Key")
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:84]: self.logger.info

**Review by mpgn — CHANGES_REQUESTED** [review]

**Comment by mpgn** [comment]
I request changes so it only print the command result

**Comment by XiaoliChan** [comment]
> I request changes so it only print the command result

Fine, will fix it tomorrow.

Any other issues?

**Comment by mpgn** [comment]
Everything works perfectly for the rest :)

File: cme/protocols/wmi/wmiexec_regout.py (hunk 1 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 2 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 4 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 5 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 6 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 3 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 7 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

**Comment by XiaoliChan** [comment]
@mpgn Please review again

**Review by mpgn — APPROVED** [review]

File: cme/protocols/wmi/wmiexec_regout.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        # Get command results
+        self.queryRegistry(keyName)
+
+    def queryRegistry(self, keyName):
+        #namespace = '//%s/root/default' % self.address
+        try:
+            descriptor, _ = self.iWbemServices.GetObject('StdRegProv')
+            descriptor = descriptor.SpawnInstance()
+            retVal = descriptor.GetStringValue(2147483650,'SOFTWARE\\classes\\hello', keyName)
+            self.logger.success(f"[+] Executed command: {self.command}")
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:98]: the [+] should be removed 
image

**Review by mpgn — CHANGES_REQUESTED** [review]

File: cme/protocols/wmi/wmiexec_regout.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            result = base64.b64decode(retVal.sValue).decode('utf-16le').rstrip('\r\n')
+            self.logger.highlight(result)
+        except:
+            self.logger.highlight("Execute command failed, probabaly got detection by AV.")
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:102]: self.logger.fail

**Review by mpgn — CHANGES_REQUESTED** [review]

File: cme/protocols/wmi/wmiexec_regout.py (hunk 8 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

**Review by mpgn — APPROVED** [review]

File: cme/protocols/wmi/wmiexec_regout.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    def execute_remote(self, data):
+        self.command = data
+        # Save result as txt file
+        self.logger.info("Executing command: \" %s \""%data)
+        resultTXT = "C:\\windows\\temp\\" + str(uuid.uuid4()) + ".txt"
+        data = data + " > " + resultTXT
+        command = self.__shell + self.encodeCommand(data)
+        self.__win32Process.Create(command, self.__pwd, None)
+        self.logger.info("[+] Waiting {}s for command completely executed.".format(self.interval_time))
```
Comment (NeffIsBack) [inline cme/protocols/wmi/wmiexec_regout.py:62]: This is probably outputting `[+]` twice now

File: cme/protocols/wmi/wmiexec_regout.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        time.sleep(self.interval_time)
+        
+        # Convert result to base64 strings
+        self.logger.info("[+] Save file to: " + resultTXT)
```
Comment (NeffIsBack) [inline cme/protocols/wmi/wmiexec_regout.py:66]: Same here and a few other in this file

File: cme/protocols/wmi/wmiexec_regout.py (hunk 10 above)
Comment (NeffIsBack) [reply cme/protocols/wmi/wmiexec_regout.py]: All good :) Thanks for the contribution

Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 11 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: Fixed

File: cme/protocols/wmi/wmiexec_regout.py (hunk 10 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py]: > All good :) Thanks for the contribution

Thanks for you & @mpgn reviews. ❤

**Review by mpgn — APPROVED** [review]

**Comment by NeffIsBack** [comment]
There are a lot of `%s` in the code. I think format strings is the new way to go, it would be nice if you could switch to it. I think there broke something because of `%s` once... just not worth it.

(If you don't know the syntax its `print(f"The user {user_name} is logged into system {system_name}.`)

**Comment by XiaoliChan** [comment]
> 

I will fix it later

**Comment by XiaoliChan** [comment]
@mpgn @NeffIsBack I rewrite the command execution module, now it is using WMI event to execute the command, no more `win32_process` (T1546.003)

**Comment by mpgn** [comment]
image
image
image

**Comment by XiaoliChan** [comment]
@mpgn Awesome looks like it works prefect. I decide to add more info output in debug.

**Comment by mpgn** [comment]
There is a problem with the output of the wmi command

image

**Comment by XiaoliChan** [comment]
> There is a problem with the output of the wmi command
> 
> image

I know this problem, I try to split out that, but I got something weird, the result will be this
```
192.168.85.200 135 Host-A Result-AAAAAA-line1
192.168.85.201 135 Host-B Result-BBBBBBB-line1
192.168.85.200 135 Host-A Result-AAAAAA-line2
192.168.85.201 135 Host-B Result-BBBBBBB-line2
```

**Comment by mpgn** [comment]
> I know this problem, I try to split out that, but I got something weird, the result will be this
192.168.85.200 135 Host-A Result-AAAAAA-line1
192.168.85.201 135 Host-B Result-BBBBBBB-line1
192.168.85.200 135 Host-A Result-AAAAAA-line2
192.168.85.201 135 Host-B Result-BBBBBBB-line2

It is exptected, it is because cme is threaded.
Otherwise you will not know which output is for which host

**Comment by XiaoliChan** [comment]
> It is exptected, it is because cme is threaded. Otherwise you will not know which output is for which host
> 
> it is like this only if you run a comand over multiple hosts and the command result takes several lines

That why I did not split out the result

**Comment by mpgn** [comment]
can you put the result of a command with the hostname etc info so I can test it ?

**Comment by XiaoliChan** [comment]
@mpgn I make a commit

**Comment by mpgn** [comment]
That is a problem on winrm I guess ^^

**Comment by mpgn** [comment]
The file with the command result is never deleted on the victim's computer

image

**Comment by XiaoliChan** [comment]
> The file with the command result is never deleted on the victim's computer
> 
> image

Well, the VBScript needs to improve.

**Comment by mpgn** [comment]
Nop

**Comment by XiaoliChan** [comment]
But if you want to clean up the temp file after the command executed, that may increase the time.

I think adding a '-clean' option is a good choice.

**Comment by mpgn** [comment]
It should remove the file automaticaly, even if it's take more time

**Comment by NeffIsBack** [comment]
What is the problem with removing the file?

**Comment by XiaoliChan** [comment]
> What is the problem with removing the file?

Probably I'm thinking wrong.

**Comment by XiaoliChan** [comment]
@mpgn Can you test this commit? it should do remove the file after the command executed [4c8cf57c4381b4a3c6229f00bf2d0669b6228883](https://github.com/XiaoliChan/CrackMapExec/commit/4c8cf57c4381b4a3c6229f00bf2d0669b6228883)

**Comment by mpgn** [comment]
The file on the Temp folder is still present with the result of the command on my side

**Comment by NeffIsBack** [comment]
> The file on the Temp folder is still present with the result of the command on my side

Did you checkout his commit?
> 4c8cf57c4381b4a3c6229f00bf2d0669b6228883

Not sure where this commit is located, but I see something like "filesystemobject delete file" in the commit changes

**Comment by XiaoliChan** [comment]
weird, it the result file should be remove after that commit.

**Comment by XiaoliChan** [comment]
@mpgn I just force push the commit, please test it again https://github.com/mpgn/CrackMapExec/commit/4c8cf57c4381b4a3c6229f00bf2d0669b6228883

**Comment by XiaoliChan** [comment]
> Not sure where this commit is located, but I see something like "filesystemobject delete file" in the commit changes

Yes, now is doing file removal after command results load into wmi instance.

**Comment by XiaoliChan** [comment]
I test that commit, and it works

Executed command
image

The temporary file was cleaned
image

**Pull request closed without merging**